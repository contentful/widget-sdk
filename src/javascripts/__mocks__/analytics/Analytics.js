export const track = jest.fn();
export const enable = jest.fn();
