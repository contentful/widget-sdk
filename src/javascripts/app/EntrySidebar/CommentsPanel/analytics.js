import * as Intercom from 'services/intercom';

const INTERCOM_PREFIX = 'feature-comments';

export function trackCommentCreated() {
  Intercom.trackEvent(`${INTERCOM_PREFIX}-comment-created`);
}
