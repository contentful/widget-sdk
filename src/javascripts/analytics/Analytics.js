import * as Config from 'Config';
import * as Snowplow from 'analytics/snowplow';
import stringifySafe from 'json-stringify-safe';
import { prepareUserData } from 'analytics/UserData';
import _ from 'lodash';
import segment from 'analytics/segment';
import { eventExists, transformEvent } from 'analytics/transform';
import * as logger from 'services/logger';
import * as analyticsConsole from 'analytics/analyticsConsole';
import * as random from '../utils/Random';
import { clearSequenceContext, initSequenceContext } from './sequenceContext';

function removeCircularRefs(obj) {
  return JSON.parse(stringifySafe(obj));
}

/**
 * @ngdoc service
 * @name analytics
 * @description
 * This service exposes an API for event tracking.
 *
 * Call to `enable` enables tracking and initializes
 * session data with user's details. There are three
 * tracking methods that are also used to collect
 * session data (`trackContextChange`, `trackStateChange`)
 * . Session data can be obtained with `getSessionData`.
 *
 * The rest of tracking is realised with calls to
 * `track` method.
 *
 * Calling `enable` doesn't mean that the events
 * will be sent to Segment and Snowplow automatically. We perform
 * an environment check to determine if we should do
 * networking.
 *
 * Once disabled, this service cannot be enabled
 * again.
 */

const ANALYTICS_ENVS = ['production', 'staging', 'preview'];
const VALUE_UNKNOWN = {};

let env = Config.env;
const session = {};
let isEnabled = false;

// Ugly but it's super tricky to simulate environment.
// Better ideas needed.
export const __testOnlySetEnv = (_env) => {
  env = _env;
};

/**
 * @ngdoc method
 * @name analytics#enable
 * @description
 * Starts event tracking
 */
export const enable = _.once((user, segmentLoadOptions) => {
  if (isEnabled) {
    return;
  }

  isEnabled = true;

  if (ANALYTICS_ENVS.includes(env)) {
    segment.enable(segmentLoadOptions);
    Snowplow.enable();
  }

  identify(prepareUserData(removeCircularRefs(user)));
  track('global:app_loaded');
});

/**
 * @ngdoc method
 * @name analytics#getSessionData
 * @param {string|array?} path
 * @param {any?} defaultValue
 * @returns {object}
 * @description
 * Gets session data. If `path` is provided then
 * extract specific nested value.
 */
export function getSessionData(path, defaultValue) {
  return _.get(session, path || [], defaultValue);
}

/**
 * @ngdoc method
 * @name analytics#track
 * @param {string} event
 * @param {object?} data
 * @description
 * Sends tracking event (with optionally provided data) to Segment and Snowplow
 * if it is on the valid events list.
 */

export function track(event, data) {
  if (!isEnabled) {
    return;
  }

  if (!eventExists(event)) {
    track('tracking:invalid_event', { event });
    return;
  }

  try {
    data = _.isObject(data) ? _.cloneDeep(data) : {};
    data = removeCircularRefs(Object.assign({}, getBasicPayload(), data));

    const transformedData = transformEvent(event, data);

    segment.track(event, transformedData);
    Snowplow.track(event, transformedData);
    analyticsConsole.add(event, transformedData);
    logEventPayloadSize(event, transformedData);
  } catch (error) {
    // ensure no errors caused by analytics will break business logic
    logger.logError('Unexpected error during event tracking', {
      error,
      message: error.message,
      event,
      data,
    });
  }
}

/**
 * This method helps to identify misuse of the analytics module
 * when SDK Value Objects sent in the event payload instead of
 * Data Transfer Objects. That results in the performance degradations
 * of the web app caused by heavy payload serialization in wootric (loaded in segment)
 *
 */
function logEventPayloadSize(event, safePayload) {
  if (typeof window.requestIdleCallback !== 'undefined') {
    window.requestIdleCallback(() => {
      try {
        const size = JSON.stringify(safePayload).length;

        // any of the payload fields has methods on the first level
        const hasMethods = Object.entries(safePayload || {})
          .flatMap(([_, v]) => Object.values(v || {}))
          .some((v) => _.isFunction(v));

        if (size > 5000 || hasMethods) {
          logger.logWarn('Potentially bloated tracking event payload', {
            event,
            size,
            hasMethods,
          });
        }
      } catch (error) {
        // ignore error
      }
    });
  }
}

/**
 * @description This function allows you to extend user's details
 * which will be sent to segment (and automatically to Intercom)
 * @param {object} params - object with new data
 */
export const updateUserInSegment = identify;

/**
 * @ngdoc method
 * @name analytics#identify
 * @param {object} extension
 * @description
 * Sets or extends session user data. Identifying
 * data is also set on Segment's client.
 */
function identify(extension) {
  session.user = session.user || {};
  const rawUserData = _.merge(session.user, extension || {});

  // we set up organization immediately, if it is not set up yet.
  // the reason for it – we might not have a space, but belong to an
  // organization. This happens, for example, after the user signs up
  // and when they have no spaces – so adding it to the session
  // automatically enriches all events with an organizationId.
  if (!session.organization) {
    // we've already removed all circular references
    // we default to the first org the user belongs to as the "current org".
    // this behaviour is the same as what the sidepanel does when it doesn't find a "current org"
    session.organization = _.get(session, 'user.organizationMemberships[0].organization');
  }

  // We need to remove the list of organization memberships as this array gets
  // flattened when it is passed to Intercom and creates a lot of noise
  const user = _.omitBy(rawUserData, (val) => _.isArray(val) || _.isObject(val));

  const userId = getSessionData('user.sys.id');

  if (userId && user) {
    segment.identify(userId, user);
    Snowplow.identify(userId);
  }

  sendSessionDataToConsole();
}

/**
 * @ngdoc method
 * @name analytics#trackContextChange
 * @param {object} space
 * @param {object} organization
 * @description
 * Sets or replaces session space and organization
 * data. Pass `null` when leaving context.
 *
 * `null` must be explicitly passed to unset the current
 * space/org contexts.
 */

export function trackContextChange(space, organization) {
  if (space) {
    session.space = removeCircularRefs(space);
  } else if (space === null) {
    session.space = null;
  }

  if (organization) {
    session.organization = removeCircularRefs(organization);
  } else if (organization === null) {
    session.organization = null;
  }

  sendSessionDataToConsole();
  track(space ? 'global:space_changed' : 'global:space_left');
}

/**
 * @ngdoc method
 * @name analytics#trackStateChange
 * @param {object} state
 * @param {object} params
 * @param {object} from
 * @param {object} fromParams
 * @description
 * Sets or replaces session navigation data.
 * Accepts arguments of `$stateChangeSuccess`
 * handler. Current state is set as a page on
 * the Segment's client.
 */
export function trackStateChange(state, params, from, fromParams) {
  if (!isEnabled) {
    return;
  }

  const data = (session.navigation = removeCircularRefs({
    state: state.name,
    params,
    fromState: from ? from.name : null,
    fromStateParams: fromParams || null,
  }));

  sendSessionDataToConsole();

  if (state.name === 'spaces.detail.entries.list') {
    initSequenceContext({ sequence_key: random.id() });
  } else {
    clearSequenceContext();
  }

  track('global:state_changed', data);
  segment.page(state.name, params);
}

function getBasicPayload() {
  // IMPORTANT: Do not add anything here without ensuring that it won't end up in any Snowplow
  //  events which might be considered invalid with additional properties unknown to their schema!
  return _.pickBy(
    {
      userId: getSessionData('user.sys.id', VALUE_UNKNOWN),
      spaceId: getSessionData('space.sys.id', VALUE_UNKNOWN),
      organizationId: getSessionData('organization.sys.id', VALUE_UNKNOWN),
      currentState: getSessionData('navigation.state', VALUE_UNKNOWN),
    },
    (val) => val !== VALUE_UNKNOWN
  );
}

function sendSessionDataToConsole() {
  analyticsConsole.setSessionData(session);
}
