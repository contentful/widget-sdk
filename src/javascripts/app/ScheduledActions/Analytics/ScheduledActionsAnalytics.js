import { once } from 'lodash';
import moment from 'moment-timezone';
import * as Analytics from 'analytics/Analytics';
import * as Intercom from 'services/intercom';
import ScheduledActionAction from '../ScheduledActionAction';

const AnalyticsEventName = {
  Dialog: 'global:dialog',
  CancelJob: 'jobs:cancel',
  CreateJob: 'jobs:create',
};

const IntercomEventName = {
  AlphaEligible: 'scheduled-publishing-alpha-eligible',
  CreateJob: 'scheduled-publishing-create-job',
};

const AnalyticsAction = {
  Publish: (entityType) => `${entityType}.publish`,
  Unpublish: (entityType) => `${entityType}.unpublish`,
};

/** Will become obsolete once we change action format in the api */
function formatAnalyticsAction(scheduledAction) {
  const { action, entity } = scheduledAction || {};

  const entityType = entity?.sys.linkType;

  switch (action) {
    case ScheduledActionAction.Publish:
      return AnalyticsAction.Publish(entityType);
    case ScheduledActionAction.Unpublish:
      return AnalyticsAction.Unpublish(entityType);
    default:
      throw new Error(`unsupported job action ${action}`);
  }
}

export function createDialogOpen() {
  const payload = {
    purpose: 'create',
    name: AnalyticsEventName.CreateJob,
    action: 'open',
  };

  return Analytics.track(AnalyticsEventName.Dialog, payload);
}

/**
 * Tracks closing of the scheduled actions modal
 * @param {*} wasSubmitted indicates if the modal was closed after submitted scheduled action or just a plain Cancel / Close
 */
export function createDialogClose(wasSubmitted) {
  const payload = {
    name: AnalyticsEventName.CreateJob,
    purpose: wasSubmitted ? 'submit' : 'cancel',
    action: 'close',
  };

  return Analytics.track(AnalyticsEventName.Dialog, payload);
}

export function createJob(job, scheduledForTimezone) {
  const payload = {
    action: formatAnalyticsAction(job),
    job_id: job.sys.id,
    scheduled_for: job.scheduledFor.datetime,
    entity_id: job.entity.sys.id,
    scheduled_for_timezone: scheduledForTimezone,
    local_timezone: moment.tz.guess(),
    timezone_offset: new Date().getTimezoneOffset(),
  };

  Intercom.trackEvent(IntercomEventName.CreateJob);
  return Analytics.track(AnalyticsEventName.CreateJob, payload);
}

export function cancelJob(job) {
  const payload = {
    action: formatAnalyticsAction(job),
    job_id: job.sys.id,
  };
  return Analytics.track(AnalyticsEventName.CancelJob, payload);
}

export const trackAlphaEligibilityToIntercom = once(() => {
  Intercom.trackEvent(IntercomEventName.AlphaEligible);
});
