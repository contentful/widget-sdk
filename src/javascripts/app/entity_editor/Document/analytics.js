import * as Analytics from 'analytics/Analytics';
import changedEntityFieldPaths from './changedEntityFieldPaths';
import { cmaGetEntity } from './api';
import * as logger from 'services/logger';
export async function trackEditConflict({
  spaceEndpoint,
  localEntity,
  localEntityFetchedAt,
  changedLocalEntity,
}) {
  let remoteEntity;
  try {
    const { sys } = localEntity;
    remoteEntity = await cmaGetEntity(spaceEndpoint, sys.type, sys.id);
  } catch (e) {
    if (e.code === 'NotFound') {
      // TODO: track with affordable analytics
      return;
    } else {
      // @ts-ignore
      logger.logError('Could not fetch remote entity for CmaDocument edit conflict tracking');
      return;
    }
  }
  const data = conflictEntitiesToTrackingData();
  Analytics.track('entity_editor:edit_conflict', data);
  function conflictEntitiesToTrackingData() {
    return {
      entityId: localEntity.sys.id,
      entityType: localEntity.sys.type,
      localChangesFieldPaths: changedEntityFieldPaths(
        localEntity.fields,
        changedLocalEntity.fields
      ).map((path) => path.join(':')),
      remoteChangesSinceLocalEntityFieldPaths: changedEntityFieldPaths(
        localEntity.fields,
        remoteEntity.fields || {}
      ).map((path) => path.join(':')),
      localEntityVersion: localEntity.sys.version,
      remoteEntityVersion: remoteEntity.sys.version,
      localEntityUpdatedAtTstamp: localEntity.sys.updatedAt,
      remoteEntityUpdatedAtTstamp: remoteEntity.sys.updatedAt,
      remoteEntityUpdatedByUserId: remoteEntity.sys.updatedBy.sys.id,
      localEntityLastFetchedAtTstamp: localEntityFetchedAt.toISOString(),
      isConflictAutoResolvable: false,
      autoConflictResolutionVersion: 1,
    };
  }
}
