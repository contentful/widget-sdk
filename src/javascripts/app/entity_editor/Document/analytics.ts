import { Entity } from './types';
import * as Analytics from 'analytics/Analytics';
import changedEntityFieldPaths from './changedEntityFieldPaths';
import { EntityRepo } from 'data/CMA/EntityRepo';
import * as logger from 'services/logger';

export async function trackEditConflict({
  entityRepo,
  localEntity,
  localEntityFetchedAt,
  changedLocalEntity,
  ...options
}: {
  entityRepo: EntityRepo;
  localEntity: Entity;
  localEntityFetchedAt: Date;
  changedLocalEntity: Entity;
  remoteEntity?: Entity;
}) {
  let remoteEntity: Entity | null;
  try {
    remoteEntity =
      options.remoteEntity || (await entityRepo.get(localEntity.sys.type, localEntity.sys.id));
  } catch (e) {
    if (e.code === 'NotFound') {
      // TODO: Track deleted entity conflict with affordable analytics. move to separate function
      //  To avoid accidentally tracking `BadRequest` errors as a conflict.
      return;
    } else {
      // @ts-ignore
      logger.logError('Could not fetch remote entity for CmaDocument edit conflict tracking', e);
      return;
    }
  }

  // Do not track if there's no conflict. E.g. if we call this function due to a `BadRequest`
  // error to accommodate for possible entity deletion kind of conflict.
  if (localEntity.sys.version !== remoteEntity.sys.version) {
    const data = conflictEntitiesToTrackingData();
    Analytics.track('entity_editor:edit_conflict', data);
  }

  function conflictEntitiesToTrackingData() {
    return {
      entityId: localEntity.sys.id,
      entityType: localEntity.sys.type,
      localChangesFieldPaths: changedEntityFieldPaths(
        localEntity.fields,
        changedLocalEntity.fields
      ).map((path) => path.join(':')),
      remoteChangesSinceLocalEntityFieldPaths: changedEntityFieldPaths(
        localEntity.fields,
        remoteEntity.fields || {}
      ).map((path) => path.join(':')),
      localEntityVersion: localEntity.sys.version,
      remoteEntityVersion: remoteEntity.sys.version,
      localEntityUpdatedAtTstamp: localEntity.sys.updatedAt,
      remoteEntityUpdatedAtTstamp: remoteEntity.sys.updatedAt,
      remoteEntityUpdatedByUserId: remoteEntity.sys.updatedBy.sys.id,
      localEntityLastFetchedAtTstamp: localEntityFetchedAt.toISOString(),
      isConflictAutoResolvable: false,
      autoConflictResolutionVersion: 1,
    };
  }
}
