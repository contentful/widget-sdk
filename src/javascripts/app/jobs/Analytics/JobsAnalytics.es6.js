import * as Analytics from 'analytics/Analytics.es6';

/**
 * Exported for testing purposes. Do not use in production !!!
 */
export const EventName = {
  Dialog: 'global:dialog',
  CancelJob: 'jobs:cancel',
  CreateJob: 'jobs:create'
};

/**
 * Atm, supported only 'Entry.publish'
 * Exported for testing purposes. Do not use in production !!!
 */
export const JobAction = {
  EntryPublish: 'Entry.publish'
};

export function createDialogOpen() {
  const payload = {
    purpose: JobAction.EntryPublish,
    name: EventName.CreateJob,
    action: 'open'
  };

  return Analytics.track(EventName.Dialog, payload);
}

export function createDialogClose() {
  const payload = {
    purpose: JobAction.EntryPublish,
    name: EventName.CreateJob,
    action: 'close'
  };

  return Analytics.track(EventName.Dialog, payload);
}

export function createJob({ jobId, scheduledAt }) {
  const payload = {
    action: JobAction.EntryPublish,
    job_id: jobId,
    scheduled_for: scheduledAt,
    timezone_offset: new Date().getTimezoneOffset()
  };

  return Analytics.track(EventName.CreateJob, payload);
}

export function cancelJob({ jobId }) {
  const payload = {
    action: JobAction.EntryPublish,
    job_id: jobId
  };
  return Analytics.track(EventName.CancelJob, payload);
}
