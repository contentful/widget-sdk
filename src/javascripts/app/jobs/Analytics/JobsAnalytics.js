import { once } from 'lodash';
import moment from 'moment-timezone';
import * as Analytics from 'analytics/Analytics';
import * as Intercom from 'services/intercom';
import JobAction from '../JobAction';

const AnalyticsEventName = {
  Dialog: 'global:dialog',
  CancelJob: 'jobs:cancel',
  CreateJob: 'jobs:create'
};

const IntercomEventName = {
  AlphaEligible: 'scheduled-publishing-alpha-eligible',
  CreateJob: 'scheduled-publishing-create-job'
};

const AnalyticsJobAction = {
  EntryPublish: 'Entry.publish',
  EntryUnpublish: 'Entry.unpublish'
};

/** Will become obsolete once we change action format in the api */
function convertActionToAnalyticsFormat(jobAction) {
  switch (jobAction) {
    case JobAction.Publish:
      return AnalyticsJobAction.EntryPublish;
    case JobAction.Unpublish:
      return AnalyticsJobAction.EntryUnpublish;
    default:
      throw new Error(`unsupported job action ${jobAction}`);
  }
}

export function createDialogOpen() {
  const payload = {
    purpose: 'job.create',
    name: AnalyticsEventName.CreateJob,
    action: 'open'
  };

  return Analytics.track(AnalyticsEventName.Dialog, payload);
}

export function createDialogClose() {
  const payload = {
    name: AnalyticsEventName.CreateJob,
    action: 'close'
  };

  return Analytics.track(AnalyticsEventName.Dialog, payload);
}

export function createJob(job, scheduledForTimezone) {
  const payload = {
    action: convertActionToAnalyticsFormat(job.action),
    job_id: job.sys.id,
    scheduled_for: job.scheduledAt,
    entity_id: job.sys.entity.sys.id,
    scheduled_for_timezone: scheduledForTimezone,
    local_timezone: moment.tz.guess(),
    timezone_offset: new Date().getTimezoneOffset()
  };

  Intercom.trackEvent(IntercomEventName.CreateJob);
  return Analytics.track(AnalyticsEventName.CreateJob, payload);
}

export function cancelJob(job) {
  const payload = {
    action: convertActionToAnalyticsFormat(job.action),
    job_id: job.sys.id
  };
  return Analytics.track(AnalyticsEventName.CancelJob, payload);
}

export const trackAlphaEligibilityToIntercom = once(() => {
  Intercom.trackEvent(IntercomEventName.AlphaEligible);
});
