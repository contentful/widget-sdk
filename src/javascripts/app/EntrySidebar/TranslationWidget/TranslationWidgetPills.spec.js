import React from 'react';
import Enzyme from 'enzyme';
import SidebarEventTypes from 'app/EntrySidebar/SidebarEventTypes.es6';
import TranslationWidgetPills from './TranslationWidgetPills.es6';
import ModalLauncherMocked from 'app/common/ModalLauncher.es6';
import { TextLink, Pill } from '@contentful/forma-36-react-components';
import { track } from 'analytics/Analytics';

jest.mock('app/common/ModalLauncher.es6', () => ({
  open: jest.fn()
}));

describe('EntrySidebar/TranslationWidgetPills', () => {
  const locales = [
    { code: 'en', default: true },
    { code: 'de', default: false },
    { code: 'es', default: false },
    { code: 'ru', default: false }
  ];

  const props = {
    emitter: {
      emit: jest.fn()
    },
    localeData: {
      activeLocales: locales,
      privateLocales: locales,
      isLocaleActive: ({ code }) => ['es-AR', 'ru'].includes(code)
    }
  };

  const render = () => Enzyme.shallow(<TranslationWidgetPills {...props} />);

  beforeEach(() => {
    props.emitter.emit.mockClear();
  });

  it('should match snapshot', () => {
    expect(render()).toMatchSnapshot();
  });

  describe('on locale deactivation', () => {
    beforeEach(() => {
      render()
        .find('.entity-sidebar__translation-pill')
        .at(1)
        .find(Pill)
        .prop('onClose')();
    });

    it('emits the DEACTIVATED_LOCALE event with the locale', () => {
      expect(props.emitter.emit).toHaveBeenCalledWith(
        SidebarEventTypes.DEACTIVATED_LOCALE,
        props.localeData.activeLocales[1]
      );
    });

    it('tracks the update event', () => {
      expect(track).toHaveBeenCalledWith('translation_sidebar:deselect_active_locale', {
        currentMode: 'multiple',
        previousActiveLocaleCount: 4,
        currentActiveLocaleCount: 3
      });
    });
  });

  describe('on change', () => {
    beforeEach(async () => {
      await render()
        .find(TextLink)
        .prop('onClick')();
      const [[callback]] = ModalLauncherMocked.open.mock.calls;
      const localeSelectDialog = callback({});
      const {
        props: { onUpdate }
      } = localeSelectDialog;
      onUpdate([
        { code: 'en-US', active: false },
        { code: 'es-AR', active: true },
        { code: 'ru', active: true },
        { code: 'de-DE', active: false }
      ]);
    });

    it('emits the SET_ACTIVE_LOCALES event with the new active locales', () => {
      expect(props.emitter.emit).toHaveBeenCalledWith(SidebarEventTypes.SET_ACTIVE_LOCALES, [
        { code: 'es-AR', active: true },
        { code: 'ru', active: true }
      ]);
    });

    it('tracks the update event', () => {
      expect(track).toHaveBeenCalledWith('translation_sidebar:update_active_locales', {
        currentMode: 'multiple',
        previousActiveLocaleCount: 4,
        currentActiveLocaleCount: 2
      });
    });
  });
});
