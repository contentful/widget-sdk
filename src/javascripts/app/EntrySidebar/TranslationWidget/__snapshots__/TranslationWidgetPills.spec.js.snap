// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntrySidebar/TranslationWidgetPills should match snapshot 1`] = `
<div
  className="pill-list entity-sidebar__translation-pills"
>
  <div
    className="entity-sidebar__translation-pill x--default"
    key="en"
  >
    <Pill
      label="en"
      status="default"
      testId="deactivate-translation"
    />
  </div>
  <div
    className="entity-sidebar__translation-pill"
    key="de"
  >
    <Pill
      label="de"
      onClose={[Function]}
      status="default"
      testId="deactivate-translation"
    />
  </div>
  <div
    className="entity-sidebar__translation-pill"
    key="es"
  >
    <Pill
      label="es"
      onClose={[Function]}
      status="default"
      testId="deactivate-translation"
    />
  </div>
  <div
    className="entity-sidebar__translation-pill"
    key="ru"
  >
    <Pill
      label="ru"
      onClose={[Function]}
      status="default"
      testId="deactivate-translation"
    />
  </div>
  <TextLink
    onClick={[Function]}
    testId="change-translation"
  >
    Change
  </TextLink>
</div>
`;
